lib/ukstore/include/uk/store.h

```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
+ *                disabled entries are indicated by the return code of
+ *                get_value()
+ * @param cookie  caller defined data. This is passed to the getter function
+ * @return        pointer to the allocated object, errptr otherwise
  */
-#define uk_store_get_entry(libname, foldername, entryname)	\
-	_uk_store_get_entry(uk_libid(libname), foldername, entryname)
+struct uk_store_object *
+uk_store_allocate_object(struct uk_alloc *a, __u64 id, const char *name,
+			 const struct uk_store_entry *entries[], void *cookie);
 
 /**
- * Decreases the refcount. When it reaches 0, the memory is freed
+ * Associates an object to the caller library
  *
  * Guard this with `#if CONFIG_LIBUKSTORE`
  *
- * @param entry the entry to release
+ * @param object     the object to add to the caller library
+ * @return           0 on success and < 0 on failure
  */
-#define uk_store_release_entry(entry)	\
-	_uk_store_release_entry((entry))
-
-void
-_uk_store_release_entry(const struct uk_store_entry *entry);
+#define uk_store_add_object(obj)					\
+	_uk_store_add_object(uk_libid_self(), (obj))
 
 int
-_uk_store_get_u8(const struct uk_store_entry *e, __u8 *out);
-int
-_uk_store_get_s8(const struct uk_store_entry *e, __s8 *out);
-int
-_uk_store_get_u16(const struct uk_store_entry *e, __u16 *out);
-int
-_uk_store_get_s16(const struct uk_store_entry *e, __s16 *out);
-int
-_uk_store_get_u32(const struct uk_store_entry *e, __u32 *out);
-int
-_uk_store_get_s32(const struct uk_store_entry *e, __s32 *out);
-int
-_uk_store_get_u64(const struct uk_store_entry *e, __u64 *out);
-int
-_uk_store_get_s64(const struct uk_store_entry *e, __s64 *out);
-int
-_uk_store_get_uptr(const struct uk_store_entry *e, __uptr *out);
-int
-_uk_store_get_charp(const struct uk_store_entry *e, char **out);
+_uk_store_add_object(__u16 library_id, struct uk_store_object *object);
+
+/**
+ * Acquires an object
+ *
+ * Increments the object's refcount and returns the object.
+ * Evern call must be paired with a call to uk_store_release_object()
+ * to decrement the refcount.
+ *
+ * @param library_id owner library id as returned by uklibid
+ * @param object_id  object id
+ * @return           pointer to object
+ */
+struct uk_store_object *
+uk_store_acquire_object(__u16 library_id, __u64 object_id);
+
+/**
+ * Releases an object
+ *
+ * Decrements the object's refcount.
+ *
+ * @param library_id owner library id as returned by uklibid
+ * @param object_id  object id
+ */
+void uk_store_release_object(__u16 library_id, __u64 object_id);
+
+/**
+ * Looks up an entry in an object of a library.
+ *
+ * @param libid    owner library (ignored for dynamic entries)
+ * @param object   target object (NULL for static entries)
+ * @param entry_id the id of the entry to look up
+ * @return the matched entry or NULL
+ */
+const struct uk_store_entry *
+uk_store_get_entry(__u16 libid, struct uk_store_object *object_id, __u64 entry_id);
```

Comment:
I remember we spoke about this function and I am wonder now if we should split this into two calls, like:
```c
const struct uk_store_entry *
uk_store_get_entry(__u16 libid, __u64 entry_id);
```
```c
const struct uk_store_object_entry *
uk_store_get_entry(struct uk_store_object *object_id, __u64 entry_id);
```
What do you think?

Reply:
Saw this comment just now, it belongs to #939 (this one is about netdev but I included the commits of #939 as a dependency). Will look at it tomorrow.